omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+
+    if (nextDtype =="UINT" || nextDtype =="INT" || nextDtype =="WORD" || nextDtype =="BOOL")
+      nextSize=2;
+    else if (nextDtype =="UDINT" || nextDtype =="DINT" || nextDtype =="REAL" || nextDtype =="DWORD")
+      nextSize=4;
+    else if (nextDtype =="ULINT" || nextDtype =="LINT" || nextDtype == "LREAL" || nextDtype =="LWORD")
+      nextSize=8;
+
+    offSetsCounted++;
+    lastOffset = newRow.back();
+    if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL" || dtype =="UDINT" || dtype =="DINT" || dtype =="REAL" || dtype =="DWORD" || dtype =="ULINT" || dtype =="LINT" || dtype == "LREAL" || dtype =="LWORD") {
+      thisOffset = lastOffset+dtypeSize;
+      if (nextSize == dtypeSize) {
+        consequitiveDtypes++;
+        if (dtype =="BOOL"){
+          consequitiveBools++;
+        }
+      }
+      else if (nextSize!=dtypeSize) {
+      /*Additional PLC padding rules mean that data must be organised into 32-bit chunks, so if a 8-bit value is included in a structure followed
+        by a different type, 24 bits of padding would be added to seperate them. Bools are even more complicated, they normally take up 16-bits, if there
+        is a single bool followed by a different datatype it will take up 32-bits. If there are 2 or more consequtive bools they will be grouped into
+        a 64-bit array with as much padding as is required. If there are 5 bools in a row, then they will take up 64 bits for the first 4 bools and then 32
+        bits for the second bool which includes 16 bits of padding*/
+        /* We compare the current and previous dtypes and if they are different then we apply padding to thisOffset, which is actually 
+        the position of the next dtype, and the last offset which is the position of the current dtype*/
+        if (dtype =="BOOL"){
+          consequitiveBools++;
+          if (consequitiveBools==1) {
+            thisOffset += 2;
+          }
+          else {
+            if ((consequitiveBools*dtypeSize)>=8) {
+              thisOffset += (consequitiveBools*dtypeSize)%8;
+            }
+            else {
+              thisOffset += 8 - consequitiveBools*dtypeSize;
+            }
+          }
+        }
+        else {
+          consequitiveBools=0;
+          if ((consequitiveDtypes*dtypeSize)>=4) {
+            thisOffset += (consequitiveDtypes*dtypeSize)%4;
+          }
+          else {
+            thisOffset += 4 - consequitiveDtypes*dtypeSize;
+          }
+        }
+        consequitiveDtypes = 1;
+      }
+    }
+    else
+    {
+      if (dtype.substr(0,7) == "STRING[")
+      {
+        auto ss = dtype.substr(7,dtype.size()-7);
+        size_t strLength = 0;
+        bool intFound = false;
+        for (size_t i = 0; i<ss.size(); i++)
+        {
+          if (ss.substr(i,1)== "]")
+          {
+            intFound = true;
+            try
+            {
+              strLength = std::stoi(ss.substr(0,i+1));
+              if (strLength < 0) throw 1;
```

Comment:
I think this is the only place where we throw and exception (and the exception is just 1???)
For the rest of this function, errors are indicated by returning -1. Its best to be consistent in the error-handling - definitely within the same function, within the class and ideally within the module, although there may be ... exceptions (pun fully intended!) to this rule at a module level.

Reply:
I dont think it really matters here what I throw, we don't print the thrown exception, rather printing the same sentence for any of the three possible causes. The main purpose of this try catch is to catch exceptions from stoi, by throwing on the if (arrayStart ... line, I save some code rather than having another flag to check. I could expand the logging by checking the exception and acting upon it, but i dont think it is necessary here.

 I can throw -1 instead though I suppose